packages/jest-cli/src/ReporterDispatcher.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
+ */
+
+'use strict';
+
+import type {RunnerContext} from 'types/Reporters';
+import type {HasteFS} from 'types/HasteMap';
+import type {Config, Path} from 'types/Config';
+import type {
+  TestResult, AggregatedResult,
+} from 'types/TestResult';
+
+export type RunOptions = {
+  estimatedTime: number,
+  showStatus: boolean,
+}
+
+class ReporterDispatcher {
+  _disabled: boolean;
+  _reporters: Array<Object>;
+  _runnerContext: RunnerContext;
+  _requiredMethods: Array<string>;
+
+  constructor(hasteFS: HasteFS, getTestSummary: () => string) {
+    this._runnerContext = {getTestSummary, hasteFS};
+    this._reporters = [];
+
+    this._requiredMethods = [];
+  }
+
+  register(reporter: Object): void {
+    if (this._validateReporter(reporter)) {
+      this._reporters.push(reporter);
+    }
+  }
+
+  unregister(ReporterClass: Function): void {
+    this._reporters = this._reporters.filter(
+      reporter => !(reporter instanceof ReporterClass),
+    );
+  }
+
+  onTestResult(
+    config: Config,
+    testResult: TestResult, 
+    results: AggregatedResult
+  ) {
+    this._callReporterMethod('onTestResult', [
+      config,
+      testResult,
+      results,
+      this._runnerContext,
+    ]);
+  }
+
+  onTestStart(config: Config, path: Path) {
+    this._callReporterMethod('onTestStart', [
+      config,
+      path,
+      this._runnerContext,
+    ]);
+  }
+
+  onRunStart(config: Config, results: AggregatedResult, options: RunOptions) {
+    this._callReporterMethod('onRunStart', [
+      config,
+      results,
+      this._runnerContext,
+      options,
+    ]);
+  }
+
+  onRunComplete(config: Config, results: AggregatedResult) {
+    this._callReporterMethod('onRunComplete', [
+      config,
+      results,
+      this._runnerContext,
+    ]);
+  }
+
+  /**
+   * Helper mehtod to call only the methods that exist
+   * on a given reporter
+   *
+   * @private
+   * @param {string} method name of the mehtod to be called
+   * @param {Array<any>} reporterArgs arguments passed in to call the reporter
+   */
+  _callReporterMethod(method: string, reporterArgs: Array<any>) {
+    this._reporters.forEach(reporter => {
+      if (reporter[method]) {
+        reporter[method](...reporterArgs);
+      }
+    });
+  }
+
+  /**
+   * _validateReporter
+   * Validates the reporters to be added by checking for the required 
+   * methods
+   * 
+   * @private
+   * @param   {Object} reporter reporter to be validated
+   * @returns {boolean} returns true if the reporter is validated
+   */
+  _validateReporter(reporter: Object) {
```

Comment:
@thymikee Oh yes, I can't see much JSDOC used within the application anyways. Removing it will make the application better.